src/main/java/io/aiven/kafka/connect/opensearch/spi/ClientsConfiguratorProvider.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 Aiven Oy
+ * Copyright 2016 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.opensearch.spi;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
+import io.aiven.kafka.connect.opensearch.OpensearchSinkConnectorConfig;
+
+public final class ClientsConfiguratorProvider {
+    private ClientsConfiguratorProvider() {
+    }
+
+    /**
+     * Use {@link ServiceLoader} mechanism to discover available configurators for Opensearch and Kafka clients 
+     * which are applicable to the provided configuration. 
+     * @param config provided configuration
+     * @return the list of discovered {@link OpensearchClientConfigurator} configurators which are applicable to the 
+     *     provided configuration.
+     */
+    public static Collection<OpensearchClientConfigurator> forOpensearch(final OpensearchSinkConnectorConfig config) {
+        final Collection<OpensearchClientConfigurator> configurators = new ArrayList<>();
+        final ServiceLoader<OpensearchClientConfigurator> loaders = ServiceLoader
+            .load(OpensearchClientConfigurator.class);
+
+        final Iterator<OpensearchClientConfigurator> iterator = loaders.iterator();
+        while (iterator.hasNext()) {
+            final OpensearchClientConfigurator configurator = iterator.next();
+            if (configurator.appliesTo(config)) {
+                configurators.add(configurator);
+            }
```

Comment:
In theory since it could be third party should not it be wrapped with `try` ... `catch`?

Reply:
Thanks @snuyanzin. This is fine I think, if we could not apply the desired configuration, it is better to fail fast on startup instead of later on during the ingestion. The example, if auth settings could not be applied, every single request will wail with `401` / `403`, what do you think?

Reply:
May be... i'm not familiar with the use case. so consider these comments as clarification comments =)
Does not it check for availability of registered configurator before starting ingestion?

Reply:
> May be... i'm not familiar with the use case. so consider these comments as clarification comments =) Does not it check for availability of registered configurator before starting ingestion?

This one I didn't get, what it means `availability of registered configurator `?

Reply:
from my point of view or how it works e.g. in Flink
1. it tries to load available connectors/formats and register them e.g. by putting in whatever collection
2. a user or an comes to schedule a query or a job and after parsing sql Flink determines what formats/connectors it should use to satisfy it and goes to check whether it has the required connector/format in those collections where it registered them during

Reply:
> from my point of view or how it works e.g. in Flink
> 
>     1. it tries to load available connectors/formats and register them e.g. by putting in whatever collection

Correct, but we do not load connectors/formats - we load additional configurators for clients (Opensearch in this case) if available. The visible effects of that are - client may not be able to connect to source or sink (no auth, no signatures, ...), exception would render connector unusable.

> 
>     2. a user or an app comes to schedule a query or a job and after parsing sql Flink determines what formats/connectors it should use to satisfy it and goes to check whether it has the required connector/format in those collections where it registered them during
> 
> 
> and in case it fails at startup, it's logged and it also will not allow to set up pipelines for the failed things

This is expected, no?